hdfs/client.py

```diff
@@ -76,22 +87,44 @@ def to_method(self, operation):
 
     def api_handler(client, hdfs_path, data=None, strict=True, **params):
       """Wrapper function."""
-      url = '%s%s%s' % (
-        client.url.rstrip('/'),
-        self.webhdfs_prefix,
-        quote(client.resolve(hdfs_path), '/= '),
-      )
       params['op'] = operation
       if client._proxy is not None:
         params['doas'] = client._proxy
-      return client._request(
-        method=self.method,
-        url=url,
-        data=data,
-        params=params,
-        strict=strict,
-        **self.kwargs
-      )
+
+      attempted_hosts = set()
+      while True:
+        # rotate the urls deque until a host is found, which was not yet tried.
+        # Lock the deque while doing this for thread safety
+        with client._lock:
+          while client._urls[0] in attempted_hosts:
+            client._urls.rotate(-1)
+          host = client._urls[0]
+
+        url = '%s%s%s' % (
+          host.rstrip('/'),
+          self.webhdfs_prefix,
+          quote(client.resolve(hdfs_path), '/= '),
+        )
+
+        try:
+          return client._request(
+            method=self.method,
+            url=url,
+            data=data,
+            params=params,
+            strict=strict,
+            **self.kwargs
+          )
+        # The execptions below indicate that the host is either unreachable, can
+        # not respond or is in Standby mode. Thus the next host should be tried.
+        except (rq.exceptions.ReadTimeout, rq.exceptions.ConnectTimeout,
+                rq.exceptions.ConnectionError, HdfsStandbyError) as exc:
+          attempted_hosts.add(host)
+          if len(attempted_hosts) == len(client._urls):
+            if len(client._urls) > 1:
```

Comment:
Is this check required?

Reply:
When the user supplied only one url, I think this message would lead to confusion, as it is not relevant. In this case it seems better to me to just reraise the error without any comments.

Reply:
Rather than delay the no URLs error until the user makes a request, let's add a message in the constructor then. For example after declaring `self._urls`:

```python
if not self._urls:
  raise HdfsError('Missing HDFS client URL')
```

Reply:
I don't quite understand your suggestion. The check here is not whether `client._urls` is empty but whether is has more than one entry. The idea here is to emit the warning that all hosts have been tried only when there were **more then one host**. If there is one host only, the warning is not needed I think.